<audio>, <video>, and soundcloud support

meandavejustice opened the pull request:
### changed
* add support for <audio> tags with direct links to audio files (via context menu)
* add support for <video> tags with direct links to video files (via context menu)
* add support for soundcloud tracks (via context menu)
* convert all components to es6 classes
* refactor to using a global event emitter for handling state changes
* pull all logic for individual players into `/client-lib/ctrls`
* move `formatTime` method into it's own file
* move `playerControls` into their own component
* move `progress` into it's own component
* move `replayView` into it's own component
* added `lib/get-soundcloud-url`
* added `lib/get-player` to determine which player is used based on the passed url
* added `audio-fft@0.1.3` dependency to help with visualizations
* added `audiosource@2.0.5` dependency to help with audio playback

- fixes #443
- fixes #442


### testing
sample frequency view
![min-vid-playing-frequency-vis](https://cloud.githubusercontent.com/assets/1844554/21152918/5de1f4be-c136-11e6-8508-458d185caac1.png)
the player/visualization should scale in all directions(I'm working on making the resolution better for all screens and sizes)
![min-vid-playing-frequency-wide](https://cloud.githubusercontent.com/assets/1844554/21152917/5dded96e-c136-11e6-94ca-3964215543da.png)
I've changed the progress bar to peek at the bottom when there is no focus on the player so the user can
have an intrusive view of their place in the media's playback
![min-vid-playing-progress-peek](https://cloud.githubusercontent.com/assets/1844554/21152919/5de27358-c136-11e6-86ca-d319a914aef5.png)
sample time domain view
![min-vid-playing-time-vis](https://cloud.githubusercontent.com/assets/1844554/21152921/5de610a8-c136-11e6-9662-bc184ccb14a7.png)
User's can click on the current time to switch between time/frequency views when audio is being played back.
![min-vid-switch-vis](https://cloud.githubusercontent.com/assets/1844554/21152920/5de54d80-c136-11e6-8813-3cbd9aab4a6a.png)





* `mkdir media`
* put an mp3 called `mother-earth.mp3` in the dir
* put an mp4 called `seneca.mp4` in the media dir
* npm install && npm run dev
* open [localhost:8080](localhost:8080) and start launching links from the context menu
* when using audio, you can click the current time in the player to switch visualizations from time to frequency


This includes some pretty big refactoring so I'd like to get a lot of eyes on it.
cc/ @6a68 @chuckharmston @clouserw @lmorchard

6a68 commented:
@meandavejustice Maybe it would be good to add (very small) media files to the repo? If you put them someplace out of the way, like a `test` or `local` directory, then there wouldn't be any need to copy files around. Would be good for contributors and good for QA.

6a68 commented on app.js:
```diff
@@ -27,11 +18,22 @@ window.AppData = new Proxy(defaultData, {
       }
     } else obj[prop] = value;
+
+    if (prop === 'src') {
```
Maybe worth checking if the old 'src' matches the new 'src', and only emitting if it's actually a change

meandavejustice replied:
There shouldn't be a case where src is updated with the same value

6a68 commented on client-lib/ctrls/audio-ctrl.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  pause() {
+    // TODO(DJ) ensure that source has loaded correctly before
```
Probably good not to land unresolved TODOs. Either fix or file a bug and add a bug number to the comment

meandavejustice replied:
https://github.com/meandavejustice/audiosource/issues/5

6a68 commented on client-lib/ctrls/audio-ctrl.js:
```diff
@@ -0,0 +1,91 @@
+  remove() {
+    // disconnect nodes and stop playback
+    // remove any listeners
+    // cleanup excess nodes
```
not totally sure this comment is 100% truthful, the function calls seem a little different

meandavejustice replied:
leftover comment from when I was planning this out

6a68 commented on client-lib/ctrls/video-ctrl.js:
```diff
@@ -0,0 +1,47 @@
+
+  remove() {
+    // listeners removed in removeVideoListeners method in
+    // `components/player-view.js` via `componentWillUnmount`
```
what does this mean? does this class not need the remove method at all?

meandavejustice replied:
Yeah, removing now

6a68 commented on client-lib/ctrls/yt-ctrl.js:
```diff
@@ -0,0 +1,62 @@
+  pause() {
+    // safeguard in case pause is called when the iframe
+    // no longer exists.
+    if (window.YTPlayer.a.contentWindow) window.YTPlayer.pauseVideo();
```
if the iframe is gone, will `window.YTPlayer.a` exist?

meandavejustice replied:
yeah, this is a weird edge case where react removes it, but the reference seems to still exist on the YTPlayer object

6a68 commented on client-lib/defaults.js:
```diff
@@ -0,0 +1,19 @@
+  duration: 0,
+  progress: 0.001, // force progress element to start out empty
+  playing: false,
+  volume: '0.5',
```
I noticed in the debugger that React seems to convert this into a float. should it be set as a float here?

meandavejustice replied:
I should probably go set this to a float everywhere, I think it was an oversight at some point and I ended up converting them all to strings for consistency. Will fix

6a68 commented on client-lib/nsa.js:
```diff
@@ -0,0 +1,210 @@
+
+function step() {
+  let currentTime = 0;
+  let progress = window.AppData.progress;
```
Suggestion: the `window.AppData.foo` pattern recurs in this file a _lot_. What about wrapping the `window.AppData` proxy object in a module? This change would way simplify unit testing code that touches AppData (just override require to pass in a sinon spy), and this file would become cleaner, too:
```js
const d = require('./app-data');

// instead of let progress = window.AppData.progress, you could do:
let progress = d.progress;
```

meandavejustice replied:
mm I like that alot, not sure why I didn't think of that before :)

6a68 commented on client-lib/nsa.js:
```diff
@@ -0,0 +1,210 @@
+  sendMetricsEvent('player_view', 'init');
+
+  if (window.AppData.player === 'audio') {
+    playerMap['audio'] = new AudioCtrl(opts);
```
What exactly are the responsibilities of this class? It seems like it not only listens for metrics-worthy events (lol nsa.js) but also does critical app setup.

Since you're using an emitter, maybe just record the metrics events in this file, and have another file that does the player setup, called something suggestive like `main.js` or whatever.

6a68 replied:
Related suggestion: seems like you could remove the last couple of `window.AppData` get/set calls from the components and just use emitters there instead. Looks like just a couple of instances in the player and progress components. I think it'd make more sense to have a hard rule than to sometimes use emitters and sometimes update state directly. Plus, again, being consistent about emitter usage simplifies unit testing components: you'd just mock out the emitter library.

6a68 commented on client-lib/nsa.js:
```diff
@@ -0,0 +1,210 @@
+emitter.on('play', (opts) => {
+  sendMetricsEvent('player_view', 'play');
+  // TODO(DJ): this is a special case to handle audio replays this is
+  // a work around for a bug in the audiosource dependency
```
this TODO needs a bug number

meandavejustice replied:
https://github.com/meandavejustice/audiosource/issues/6

6a68 commented on client-lib/nsa.js:
```diff
@@ -0,0 +1,210 @@
+  // if we are paused force the ui to update
+  if (!window.AppData.playing) {
+    window.AppData = deepAssign(window.AppData, {
+      time: `${formatTime(opts.value)} / ${formatTime(window.AppData.duration)}`,
```
l10n nit: this might need to be reversed in rtl languages. maybe file a good first bug if it's not an obvious, easy fix

6a68 replied:
hmm. it might also make sense to split this time thing into 'elapsed time' and 'total time'. that would let you do the locale-specific display stuff in the display code, where it belongs, and not in the data management code.

6a68 commented on client-lib/nsa.js:
```diff
@@ -0,0 +1,210 @@
+  window.AppData.minimized = false;
+});
+
+emitter.on('send-to-tab', () => {
```
I'd suggest moving the send-to-tab listener into the sendToAddon library. keep the related stuff together.

6a68 replied:
or, at least create another listener that does the sendToAddon bit. might require another event, since it looks like currentTime gets updated just before the addon is notified.

meandavejustice replied:
created issue https://github.com/meandavejustice/min-vid/issues/526

6a68 commented on components/error-view.js:
```diff
@@ -27,14 +31,16 @@ module.exports = React.createClass({
     return (
-        <div className='error' onMouseEnter={this.enterView} onMouseLeave={this.leaveView}>
-          <div className={cn('controls', 'drag', {hidden: !this.state.hovered, minimized: this.props.minimized})}>
+        <div className='error' onMouseEnter={this.enterView.bind(this)} onMouseLeave={this.leaveView.bind(this)}>
```
I'm surprised you need `this` binding. does react not take care of this?

meandavejustice replied:
Not when using es6 classes, https://facebook.github.io/react/docs/handling-events.html

6a68 commented on components/player-view.js:
```diff
@@ -1,318 +1,137 @@
+module.exports = class Player extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {hovered: false};
```
This seems to occur everywhere. Any reason it couldn't be pulled up into a parent base class?

6a68 commented on components/player-view.js:
```diff
@@ -1,318 +1,137 @@
+        deepAssign(window.AppData, {
+          error: true
+        });
+      }, 30000); // 30 second timeout
```
I wonder if some kind of intermediate "working on it..." state should be shown after 5-10 seconds. 30 seconds is a really long time to wait on the web

meandavejustice replied:
bumped down to 20 seconds for now and opened https://github.com/meandavejustice/min-vid/issues/527

6a68 commented on components/player-view.js:
```diff
@@ -1,318 +1,137 @@
-        }
+    if (this.props.player === 'youtube') {
+      emitter.emit('init', {
+        onLoaded: this.onLoaded.bind(this)
```
Using an emitter to pass around a function reference? This seems weird to me.

Could you fire a new event instead and wire up the function directly inside the new event's listener?

meandavejustice replied:
onLoaded clears the timeout and calls the onError method, so It needs to be on the component.

meandavejustice replied:
`onLoaded` is the same regardless since it is just the method right here in `player-view` component. However, I need to be able to call it at different times depending on which `ctrl` is being used.
For instance. in `audio-ctrl` I need to load the underlying file before calling `onLoaded` in that file.

With YouTube I need to call it withing the `init` method because that is where we initialize the playerMap objects so that we can control playback in one place for all events(play, pause, etc..)

6a68 commented on components/player-view.js:
```diff
@@ -1,318 +1,137 @@
-    sendMetricsEvent('player_view', 'close');
+  onError() {
+    if (!this.props.src) return;
+    window.AppData.error = true;
```
Fire error event instead of setting AppData here?

6a68 commented on lib/get-player.js:
```diff
@@ -0,0 +1,17 @@
+function isAudio(src) {
+  const isSoundcloud = new RegExp('^(https?:)?//api.soundcloud.com\/tracks\/[0-9]+\/stream')
+        .exec(src);
+  const isAudio = new RegExp('^(https?:)?//*.+(.mp3|.opus|.weba|.ogg|.wav|.flac)')
```
Definitely should DRY up these recurring RegExps. Maybe put inside a 'supported-formats' lib or something, just so you don't have to remember to update multiple spots if you add / remove formats.

fwiw, the regexp literal is compact and pretty nice, except you'd have to escape the // :

```js
const isAudio = /^(https?:)?\/\/*.+(.mp3|.opus|.weba|.ogg|.wav|.flac)/
```

meandavejustice replied:
#525

6a68 commented on lib/get-soundcloud-url.js:
```diff
@@ -0,0 +1,23 @@
+
+function getSoundcloudUrl(videoUrl, cb) {
+  Request({
+    url: 'https://api.soundcloud.com/resolve.json?client_id=' + clientId + '&url=' + videoUrl,
```
This feels like it should be a template string constant defined on the class

6a68 commented on locales/en-US/addon.properties:
```diff
@@ -6,8 +6,11 @@ tooltip_send_to_tab= Send to tab
-loading_msg= Loading video from %s
-content_menu_msg= Send to mini player
\ No newline at end of file
+loading_msg= Loading %s from %s
```
you'll need to update the localization note to make clear which placeholder is which

6a68 commented:
@meandavejustice Aight, did a first pass, made some suggestions. I'm fine to defer most of them until later, provided a bug is filed.

I'll give the other reviewers a chance to weigh in. Ping me when you'd like me to do a final nitpicky pass after you've made changes and/or filed followup bugs.

meandavejustice commented on components/loading-view.js:
```diff
@@ -2,20 +2,24 @@ const React = require('react');
-    return {hovered: false};
-  },
-  enterView: function() {
+module.exports = class ErrorView extends React.Component {
```
Whoops

meandavejustice commented:
@6a68 regarding https://github.com/meandavejustice/min-vid/pull/524#issuecomment-266839606

I think I would rather add a task to create the `media` dir and download some files from a server. I'd really hate to bloat the repo with extra files, it could really slow down pulling/cloning on poor connections

meandavejustice commented on app.js:
```diff
@@ -27,11 +18,22 @@ window.AppData = new Proxy(defaultData, {
         window.console.error('Unable to parse l10n strings: ', ex);
       }
     } else obj[prop] = value;
```
Hmm, I don't think I'm very concerned about this type of bug, we only send data down in launch-video, i think It would be excessive to add these checks/validation

meandavejustice commented:
@6a68 You mind taking another look at this?

6a68 commented:
I just asked again in the status mtg to get another set of eyes on this. Letting it sit for a little bit to see if anyone makes time to help out.

If nothing major comes up, I'll plan to do my second & final pass this afternoon.

6a68 commented:
@meandavejustice I have a review in process; finishing it is at the top of my todo list for tomorrow. Couple questions for now:

When I right-click the 'bad' souncloud link, there's no minvid context menu option. Is that intentional, or a bug?

Also, I can't get the 'good' soundcloud test link to work. It looks like `client_id=undefined` in the first error in this screenshot:

<img width="1060" alt="screen shot 2016-12-19 at 5 02 13 pm" src="https://cloud.githubusercontent.com/assets/96396/21334504/fb4b8ba4-c60c-11e6-91d7-5ed730ff4d86.png">

Am I missing a setup step? I just did `npm install && npm run dev` with dev edition running.

meandavejustice commented:
@6a68 I've just pushed a fix for the soundcloud issue. I had mistyped the object I was accessing the client id from in `lib/get-soundcloud-url`.

As for the context menu not working on "bad soundcloud url", this is intended. The regex that checks which urls to apply the `send-to-player` option do not match. I've updated the `index.html` to add a separate option for a soundcloud track which will error on fetch, and updated the link text for the "no context menu" option.

6a68 commented on lib/context-menu-handlers.js:
```diff
@@ -17,6 +18,23 @@ const contentScriptVimeoCase = `self.on('click', function (node, data) {
+  self.postMessage({
+    src: node.href || node.src,
+    url: window.location.href,
+    domain: new URL(node.href || node.src).host
```
The `URL` constructor throws if the string isn't a valid URL. Could you wrap this in a try-catch, and not send the postmessage if it throws?

meandavejustice replied:
Good catch, will do

6a68 commented on lib/context-menu-handlers.js:
```diff
@@ -17,6 +18,23 @@ const contentScriptVimeoCase = `self.on('click', function (node, data) {
+self.on('click', function (node) {
+  self.postMessage({
+    url: node.href,
+    domain: new URL(node.href).host
```
Same thing here, wrap in a try/catch, and don't send unless it's valid.

meandavejustice replied:
👍

6a68 commented on lib/get-soundcloud-url.js:
```diff
@@ -0,0 +1,22 @@
+      }
+
+      if (resp.json.kind !== 'track' || !resp.json.streamable) {
+        cb('not track or not streamable');
```
For telling the user what went wrong, might be useful to split this error to distinguish between the 'not track' and 'not streamable' cases? Maybe a good first bug?

meandavejustice replied:
This currently only sets the error to TRUE when sent down to the client, it's more or less a placeholder until we add more specific error messaging. I've created https://github.com/meandavejustice/min-vid/issues/532 to track these types of instances

6a68 commented on package.json:
```diff
@@ -20,13 +20,16 @@
   "bugs": {
     "url": "https://github.com/meandavejustice/min-vid/issues"
   },
+  "config": {
```
Not a change you made in this PR, but the version `0.3.3b` should be `0.3.3-b` to follow semver.

meandavejustice replied:
👍

meandavejustice commented on client-lib/format-time.js:
```diff
@@ -0,0 +1,10 @@
+  const days = Math.floor(seconds / secondsInDay);
+  seconds = seconds % secondsInDay;
+  const utc = new Date(seconds * 1000);
+  return `${days}:${utc.toLocaleTimeString('en-US', {hour12: false, timeZone: 'UTC'})}`
```
https://github.com/meandavejustice/min-vid/issues/526

meandavejustice commented on client-lib/app-data.js:
```diff
@@ -0,0 +1,49 @@
+  error: false,
+  muted: false,
+  exited: false,
+  time: '0:00 / 0:00',
```
https://github.com/meandavejustice/min-vid/issues/526

meandavejustice commented:
@6a68 Thanks for your review

>Bug 1: when audio is loading, close button doesn't work.
Woah, gunna take care of this one, the underlying library that I had written a while back really needs a refresh, will be eager to do that soon. Will take care of this in our code for now.

> Bug 2: when audio is loading, send to tab doesn't close the window.
Pretty much, same as above :)

I think I've address the rest of your comments. Let's talk in the morning to see if there is anything else we should try and get in before the next release.

meandavejustice commented on components/error-view.js:
```diff
@@ -4,17 +4,21 @@ const sendToAddon = require('../client-lib/send-to-addon');
+module.exports = class ErrorView extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {hovered: false};
```
It has no effect on the actual dom's state. It's just initializing the state object with a value. The same way I would initialize a variable representing hover state in a vanilla js project.

``` js
var textHovered = false;
document.querySelector('.text').addEventListener('mouseenter', (ev) => textHovered = true);
document.querySelector('.text').addEventListener('mouseleave', (ev) => textHovered = false);
```

meandavejustice commented:
can't reply to https://github.com/meandavejustice/min-vid/pull/524#discussion_r93340515 for some reason so:
>Sure, I was more curious about why you have to manually initialized hovered to false. Is this just to clear out hover state when switching between views, like from loading to player view?

Variables for states like "hover" should never be initialized with true, since it is highly unlikely that the user is hovering over the element on initialization, and they are set to hover on the `mouseenter/over` events and removed on `mouseleave`

6a68 commented:
> Variables for states like "hover" should never be initialized with true, since it is highly unlikely that the user is hovering over the element on initialization, and they are set to hover on the mouseenter/over events and removed on mouseleave

Right, makes sense. I wasn't thinking that the data belonged only to the specific component, not to the global data store.

6a68 commented:
@meandavejustice Looks like those major bugs were fixed. Thanks for filing followups for the other little issues. Merging :tada:
